AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,130 @@
+
+  // Read token array from storage.
+  NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+  if (!tokenArray || tokenArray.count == 0) {
+    return nil;
+  }
+
+  MSAuthTokenInfo *authTokenInfo = tokenArray.firstObject;
+  NSDate *nextChangeTime = tokenArray.count > 1 ? tokenArray[1].startTime : nil;
+  if ([authTokenInfo.endTime laterDate:nextChangeTime]) {
+    return [[MSAuthTokenInfo alloc] initWithAuthToken:authTokenInfo.authToken
+                                         andStartTime:authTokenInfo.startTime
+                                           andEndTime:nextChangeTime];
+  }
+  return authTokenInfo;
+}
+
+- (void)saveAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+    if (!tokenArray) {
+      tokenArray = [NSMutableArray<MSAuthTokenInfo *> new];
+    }
+    if (tokenArray.count == 0) {
+
+      // Add nil token if the entire story is empty.
+      [tokenArray addObject:[MSAuthTokenInfo new]];
+    }
+    if (![tokenArray.lastObject.authToken isEqual:authToken]) {
+
+      // If new token differs from the last token of array - add it to array.
+      MSAuthTokenInfo *newAuthToken = [[MSAuthTokenInfo alloc] initWithAuthToken:authToken andStartTime:[NSDate date] andEndTime:expiresOn];
+      [tokenArray addObject:newAuthToken];
+    }
+
+    // Cap array size at max available size const (deleting from beginning).
+    if ([tokenArray count] > kMSIdentityMaxAuthTokenArraySize) {
+      [tokenArray removeObjectAtIndex:0];
+    }
+
+    // Save new array.
+    if ([MSKeychainUtil storeArray:tokenArray forKey:kMSIdentityAuthTokenArrayKey]) {
+      MSLogDebug([MSIdentity logTag], @"Saved new history state in keychain.");
+    } else {
+      MSLogWarning([MSIdentity logTag], @"Failed to save new history state in keychain.");
+    }
+    if (authToken) {
+      [self saveTokenToKeychain:authToken];
+    } else {
+      [self deleteTokenFromKeychain];
+    }
+    if (authToken && accountId) {
+      [MS_USER_DEFAULTS setObject:(NSString *)accountId forKey:kMSIdentityMSALAccountHomeAccountKey];
+    } else {
+      [MS_USER_DEFAULTS removeObjectForKey:kMSIdentityMSALAccountHomeAccountKey];
+    }
+  }
+}
+
+- (void)removeAuthToken:(nullable NSString *)authToken {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+
+    // Do nothing if there's just one entry in the history or no history at all.
+    if (!tokenArray || tokenArray.count == 1) {
+      return;
+    }
+
+    // Find, delete the oldest entry. Do not delete the most recent entry.
+    for (NSUInteger i = 0; i < tokenArray.count - 1; i++) {
```

Comment:
Please allow only the oldest token to be removed.

Reply:
Sure. But I'd rather add this check to the PR #1418 about tokens deletion, not into this PR. In this PR `removeAuthToken` method is unused and we only changed tokens storing logic. I think it would be better to apply our changes iteratively and finish the implementation of tokens removal as part of the separate story. I'll add a `TODO` so we don't forget to implement this.

I think we shouldn't have implemented tokens removal as part of this PR because it makes PR too huge and hard to review. But since we already implemented some removal logic here then I'd better keep it as is and fix in the next PR.

Reply:
@MatkovIvan why would we have a parameter to search authToken then, if we only need to delete the oldest token? I think it might be good to refactor this to always delete only the oldest auth token, without parameters.

Reply:
Already discussed. See the comment for this method.